maintenance/troubleshoot/vpp/tracing-packets.md

```diff
@@ -0,0 +1,272 @@
+````
+vpp# pcap dispatch trace on max 1000 file vppcapture buffer-trace dpdk-input 1000
+vpp# pcap dispatch trace off
+````
+This will generate a file named `/tmp/vppcapture`
+
+Then on your host run:
+````bash
+calivppctl sh vpp NODENAME
+root@server:~# mv /tmp/vppcapture /run/vpp/
+root@server:~# exit
+# The file should now be at /run/vpp/vppcapture on your host 'NODENAME'
+````
+
+You can then `scp vppcapture` to your machine and open it with wireshark
+[More info about this here](https://haryachyy.wordpress.com/2019/09/29/learning-vpp-trace-with-wireshark/)
+
+### Traffic received in the pods
+
+In order to inspect traffic actually received by the pods if you have `tcpdump` installed in the pod and simply `tcpdump -ni eth0` inside the pod.
+
+#### Tcpdump only available on the host
+
+* provided that you have `tcpdump` installed on the host, you could use nsenter to attach to the pod's network namespace and use the host's `tcpdump` against the container's interface.
+
+This works on docker as follows :
+````bash
+{% raw %}
+# Find the container ID you want to inspect
+$ docker ps
+CONTAINER ID        IMAGE                        COMMAND                CREATED              STATUS              PORTS               NAMES
+4c01db0b339c        ubuntu:12.04                 bash                   17 seconds ago       Up 16 seconds       3300-3310/tcp       webapp
+
+# Get the container PID out of it
+$ docker inspect --format '{{ .State.Pid }}' 4c01db0b339c
+12345
+
+# Attach
+$ sudo nsenter -t 12345 -n bash
+{% endraw %}
+````
+
+#### No tcpdump, but we have python !
+
+Open an AF_PACKET socket in python with the following code
+and run it attached to the running namespace as previously.
+````python
+#!/usr/bin/env python
+from socket import *
+from struct import unpack
+
+IFNAME = "eth0"
+N_PKT = 50
+
+sock = socket(AF_PACKET, SOCK_DGRAM, 0x0800)
+sock.bind((IFNAME, 0x0800))
+for _ in range(N_PKT):
+    data = sock.recvfrom(1500, 0)[0]
+    src_addr = inet_ntop(AF_INET, data[12:16])
+    dst_addr = inet_ntop(AF_INET, data[16:20])
+    src_port, = unpack("!H", data[20:22])
+    dst_port, = unpack("!H", data[22:24])
+    data_len, = unpack("!H", data[24:26])
+    cksum, = unpack("!H", data[26:28])
+
+    print("%s:%d -> %s:%d len %d cs %d" % (src_addr, src_port, dst_addr, dst_port, data_len, cksum))
+````
+
+You can use `echo "the python blob" | sudo nsenter -t <thePID> -n python`
+
+### Traffic to the k8s agent
+
+As the kubelet agent runs directly on the host without a network namespace, pods talking to it (e.g. coredns resolvers) would go through a specific path. Packets destined to it will be caught by VPP's punt mechanism, and will be forwarded to the host through a tun interface called `vpptap0`.
+
+To debug traffic within VPP, use the trace & check that traffic is correctly punted to a tun interface corresponding to `vpptap0` in linux.
+
+On the host, simply use `tcpdump -ni vpptap0`
+
+### Further documentation
```

Comment:
Above and beyond

Use "open in new window" code:

{% include open-new-window.html text='NAME' url='URL' %}